protocol/message_persistence.go

```diff
@@ -1331,6 +1448,215 @@ func (db sqlitePersistence) MessageByChatIDs(chatIDs []string, currCursor string
 	return result, newCursor, nil
 }
 
+// This is very YOLO. We probably should use a limit and cursor but let's see
+func (db sqlitePersistence) AllMessagesForBackup() ([]*protobuf.BackedUpMessage, error) {
+	// Build a new column `cursor` at the query time by having a fixed-sized clock value at the beginning
+	// concatenated with message ID. Results are sorted using this new column.
+	// This new column values can also be returned as a cursor for subsequent requests.
+	where := "WHERE NOT(m1.hide)"
+	fields := db.tableUserMessagesProtobufFields()
+	selectQuery := `SELECT    %s
+				FROM      user_messages m1
+				LEFT JOIN pin_messages pm
+				ON 	      m1.id = pm.message_id AND pm.pinned = 1`
+	query := fmt.Sprintf(selectQuery, fields) + " " + where
+	rows, err := db.db.Query(query)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	var messages []*protobuf.BackedUpMessage
+	messageIdx := make(map[string]*protobuf.BackedUpMessage, 0)
+	for rows.Next() {
+		message := &protobuf.BackedUpMessage{}
+		if err := db.tableUserMessagesScanProtobufFields(rows, message); err != nil {
+			return nil, err
+		}
+
+		if _, ok := messageIdx[message.Id]; !ok {
+			messageIdx[message.Id] = message
+			messages = append(messages, message)
+		}
+	}
+
+	return messages, nil
+}
+
+func (db sqlitePersistence) SaveBackedUpMessages(messages []*protobuf.BackedUpMessage) error {
+	if len(messages) == 0 {
+		return nil
+	}
+
+	tx, err := db.db.BeginTx(context.Background(), &sql.TxOptions{})
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err == nil {
+			err = tx.Commit()
+			return
+		}
+		// don't shadow original error
+		_ = tx.Rollback()
+	}()
+
+	allFields := db.tableUserMessagesAllFields()
+	valuesVector := strings.Repeat("?, ", db.tableUserMessagesAllFieldsCount()-1) + "?"
+	query := "INSERT OR REPLACE INTO user_messages(" + allFields + ") VALUES (" + valuesVector + ")"
+	stmt, err := tx.Prepare(query)
+	if err != nil {
+		return err
+	}
+	defer stmt.Close()
+
+	// pinMessages := make([]*common.PinMessage, 0)
+
+	for _, msg := range messages {
+		allValues, err := db.backedUpMessageToUserMessageValues(msg)
+		if err != nil {
+			return err
+		}
+
+		_, err = stmt.Exec(allValues...)
+		if err != nil {
+			return err
+		}
+
+		// if msg.PinnedBy != "" {
+		// 	pinMessage := protobuf.PinMessage{
+		// 		Clock:       msg.Timestamp,
+		// 		MessageId:   msg.Id,
+		// 		ChatId:      msg.ChatId,
+		// 		MessageType: msg.MessageType,
+		// 		Pinned:      true,
+		// 	}
+		// 	pinMessages = append(pinMessages, &common.PinMessage{
+		// 		ID:               msg.Id, // TODO do we need a special ID?
+		// 		From:             msg.PinnedBy,
+		// 		PinMessage:       &pinMessage,
+		// 		WhisperTimestamp: msg.Timestamp,
+		// 	})
+		// }
+	}
+
+	// FIXME this crashes
+	// if len(pinMessages) > 0 {
+	// 	err := db.SavePinMessages(pinMessages)
```

Comment:
I'd need help here. It crashes in `buildPinMessageQueries` when it calls `db.db.BeginTx`. I have no idea what would be wrong. I only tested in tests, so maybe the issue doesn't happen in a real app.